File: src/main/java/ru/ioleksiv/telegram/bot/api/annotations/filter/secondary/VenueFilter.java
package ru.ioleksiv.telegram.bot.api.annotations.filter.secondary;

import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.NotNullFilter;
import ru.ioleksiv.telegram.bot.api.annotations.filter.primitive.StringFilter;
import ru.ioleksiv.telegram.bot.api.model.annotation.AnnotationState;

import java.lang.annotation.Documented;
import java.lang.annotation.ElementType;
import java.lang.annotation.Retention;
import java.lang.annotation.RetentionPolicy;
import java.lang.annotation.Target;

@Retention(RetentionPolicy.RUNTIME)
@Target(ElementType.METHOD)
@Documented
public @interface VenueFilter {
    AnnotationState value() default AnnotationState.ON;

    LocationFilter location() default @LocationFilter(AnnotationState.OFF);

    StringFilter title() default @StringFilter(AnnotationState.OFF);

    StringFilter address() default @StringFilter(AnnotationState.OFF);

    NotNullFilter foursquareId() default @NotNullFilter(AnnotationState.OFF);

    StringFilter foursquareType() default @StringFilter(AnnotationState.OFF);

}

File: core/src/main/kotlin/io/github/oleksivio/tl/kbot/core/annotations/filter/composite/VenueFilter.kt
package io.github.oleksivio.tl.kbot.core.annotations.filter.composite

import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.NotNullFilter
import io.github.oleksivio.tl.kbot.core.annotations.filter.primitive.StringFilter
import io.github.oleksivio.tl.kbot.core.model.annotation.AnnotationState

@Retention
@Target(AnnotationTarget.FUNCTION, AnnotationTarget.PROPERTY_GETTER, AnnotationTarget.PROPERTY_SETTER)
@MustBeDocumented
annotation class VenueFilter(
    val status: AnnotationState = AnnotationState.ON,
    val validator: Array<String> = [],
    val location: LocationFilter = LocationFilter(
        status = AnnotationState.OFF
    ),
    val title: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val address: StringFilter = StringFilter(
        status = AnnotationState.OFF
    ),
    val foursquareId: NotNullFilter = NotNullFilter(
        status = AnnotationState.OFF
    ),
    val foursquareType: StringFilter = StringFilter(
        status = AnnotationState.OFF
    )
)
